internal/pkg/manager/recordingmerger/apparmor.go

```diff
@@ -0,0 +1,298 @@
+		return b
+	}
+	if b == nil {
+		return a
+	}
+
+	merged := newAppArmorPathSet(a)
+	for _, path := range *b {
+		if !merged.Matches(path) {
+			merged.Add(path)
+		}
+	}
+
+	return merged.Patterns()
+}
+
+func mergeFilesystem(base, additions *apparmorprofileapi.AppArmorAbstract) {
+	//nolint:nestif  // refactoring this makes it worse
+	if base.Filesystem != nil && additions.Filesystem != nil {
+		r := newAppArmorPathSet(base.Filesystem.ReadOnlyPaths)
+		w := newAppArmorPathSet(base.Filesystem.WriteOnlyPaths)
+		rw := newAppArmorPathSet(base.Filesystem.ReadWritePaths)
+
+		if additions.Filesystem.ReadWritePaths != nil {
+			for _, p := range *additions.Filesystem.ReadWritePaths {
+				if rw.Matches(p) {
+					// no changes necessary
+				} else if pat := r.PopMatching(p); pat != nil {
+					rw.Add(*pat)
+				} else if pat := w.PopMatching(p); pat != nil {
+					rw.Add(*pat)
+				} else {
+					rw.Add(p)
+				}
+			}
+		}
+		if additions.Filesystem.ReadOnlyPaths != nil {
+			for _, p := range *additions.Filesystem.ReadOnlyPaths {
+				if rw.Matches(p) {
+					// no changes necessary
+				} else if r.Matches(p) {
+					// no changes necessary
+				} else if pat := w.PopMatching(p); pat != nil {
+					rw.Add(*pat)
+				} else {
+					r.Add(p)
+				}
+			}
+		}
+		if additions.Filesystem.WriteOnlyPaths != nil {
+			for _, p := range *additions.Filesystem.WriteOnlyPaths {
+				if rw.Matches(p) {
+					// no changes necessary
+				} else if pat := r.PopMatching(p); pat != nil {
+					rw.Add(*pat)
+				} else if w.Matches(p) {
+					// no changes necessary
+				} else {
+					w.Add(p)
+				}
+			}
+		}
+		base.Filesystem = &apparmorprofileapi.AppArmorFsRules{
+			ReadOnlyPaths:  r.Patterns(),
+			WriteOnlyPaths: w.Patterns(),
+			ReadWritePaths: rw.Patterns(),
+		}
+	} else if additions.Filesystem != nil {
+		base.Filesystem = additions.Filesystem
+	}
+}
+
+func newAppArmorPathSet(patterns *[]string) appArmorPathSet {
+	m := appArmorPathSet{}
+	if patterns != nil {
+		for _, p := range *patterns {
+			m.Add(p)
+		}
+	}
+	return m
+}
+
+type appArmorPathSet struct {
+	paths []apparmorPath
+}
+
+type apparmorPath struct {
+	pattern string
+	expr    *regexp.Regexp
+}
+
+func (m *appArmorPathSet) findMatch(path string) int {
+	for i, p := range m.paths {
+		if p.pattern == path {
+			return i
+		}
+		if p.expr != nil && p.expr.MatchString(path) {
+			return i
+		}
+	}
+	return -1
+}
+
+func (m *appArmorPathSet) Matches(path string) bool {
+	return m.findMatch(path) >= 0
+}
+
+func (m *appArmorPathSet) PopMatching(path string) *string {
+	i := m.findMatch(path)
+	if i >= 0 {
+		ret := m.paths[i].pattern
+		m.paths[i] = m.paths[len(m.paths)-1]
+		m.paths = m.paths[:len(m.paths)-1]
+		return &ret
+	}
+	return nil
+}
+
+func (m *appArmorPathSet) Add(pattern string) {
+	rex, err := appArmorGlobToRegex(pattern)
+	if err != nil {
+		log.Printf("Failed to parse AppArmor glob pattern '%s': %x\n", pattern, err)
+	}
+	m.paths = append(m.paths, apparmorPath{
```

Comment:
nit: this is going to merge paths even though an error occurs, is this as expected?

Reply:
Yes. If `appArmorGlobToRegex` fails, we still want to add it to match on the raw pattern (just not on the regular expression). :)